### PR TITLE
Comparison notebook: TruLens groundedness vs RAGAS faithfulness. and add dataset preprocessing utils

### PR DESCRIPTION
**Pull request by sfc-gh-dhuang**
# Description

See doc for more experiment details: https://docs.google.com/document/d/1oleo8A3pP1CUIvKIL9B4fT4qeSaAVesFEAXKuwFFDvg/edit#heading=h.he809ei5ojer

Please include any other details of this change useful for _TruLens_ developers.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to
  not work as expected)
- [ ] New Tests
- [ ] This change includes re-generated golden test results
- [ ] This change requires a documentation update

### CODE REVIEW

File: src/benchmark/trulens/benchmark/benchmark_frameworks/experiments/dataset_preprocessing.py [hunk 1]
```diff
@@ -0,0 +1,467 @@
...
+
+import matplotlib.pyplot as plt
+import pandas as pd
+import seaborn as sns
+from sklearn.metrics import confusion_matrix
+from trulens.feedback import GroundTruthAggregator
+
+
+def generate_summeval_groundedness_golden_set(
+    file_path, max_samples_per_bucket=200
```
Comment (sfc-gh-chu) [inline src/benchmark/trulens/benchmark/benchmark_frameworks/experiments/dataset_preprocessing.py:15]: can we add typing to these methods too?

File: src/benchmark/trulens/benchmark/benchmark_frameworks/experiments/dataset_preprocessing.py [hunk 2]
```diff
@@ -0,0 +1,467 @@
...
+                    )
+
+            except json.JSONDecodeError as e:
+                print(f"JSON decode error: {e}. Check the line format.")
+
+
+# Snowflake IT dataset
+
+
+def generatate_snowflake_it_golden_set_groundedness(file_path):
```
Comment (sfc-gh-chu) [inline src/benchmark/trulens/benchmark/benchmark_frameworks/experiments/dataset_preprocessing.py:108]: typo (`generate`) here

File: src/benchmark/trulens/benchmark/benchmark_frameworks/experiments/dataset_preprocessing.py [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+    with open(file_path, "r", encoding="utf-8") as f:
+        for line in f:
+            # Parse each line as a JSON object
+            try:
+                data = json.loads(line)
+
+                # Ensure the expected keys exist in the JSON
+                try:
+                    row = data
+                    assert (
```
Comment (sfc-gh-chu) [inline src/benchmark/trulens/benchmark/benchmark_frameworks/experiments/dataset_preprocessing.py:56]: Is this method intended only for use with the notebook and summeval? Would it make sense to try to generalize this to support other schemas if we want to incorporate it in the open source package?

File: pyproject.toml [hunk 4]
```diff
@@ -6,7 +6,7 @@ requires = [
 
 [tool.poetry]
 name = "trulens"
-version = "1.0.1"
+version = "1.1.0"
```
Comment (sfc-gh-chu) [inline pyproject.toml:9]: these might be unintentional changes im guessing

File: src/benchmark/trulens/benchmark/benchmark_frameworks/experiments/competitive_analysis.ipynb [hunk 5]
```diff
@@ -0,0 +1,573 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## TruLens vs RAGAS performance analysis for groundedness or factuality\n",
```
Comment (sfc-gh-jreini) [inline src/benchmark/trulens/benchmark/benchmark_frameworks/experiments/competitive_analysis.ipynb:7]: delete "or factuality"

File: src/benchmark/trulens/benchmark/benchmark_frameworks/experiments/dataset_preprocessing.py (hunk 1 above)
Comment (sfc-gh-dhuang) [reply src/benchmark/trulens/benchmark/benchmark_frameworks/experiments/dataset_preprocessing.py]: yes done

File: src/benchmark/trulens/benchmark/benchmark_frameworks/experiments/dataset_preprocessing.py (hunk 3 above)
Comment (sfc-gh-dhuang) [reply src/benchmark/trulens/benchmark/benchmark_frameworks/experiments/dataset_preprocessing.py:58]: so for now this method is quite specific to SummEval

File: src/benchmark/trulens/benchmark/benchmark_frameworks/experiments/rag_triad_baselines.ipynb [hunk 6]
```diff
@@ -23,241 +23,29 @@
    "outputs": [],
```
Comment (sfc-gh-jreini) [inline src/benchmark/trulens/benchmark/benchmark_frameworks/experiments/rag_triad_baselines.ipynb:23]: <pre class="ql-syntax" spellcheck="false">Line #4.    ms_marco_hard_neg_balanced = pd.read_csv("data/ms_marco_hard_neg_balanced.csv")
</pre><p>can we show the BEIR loader here instead of reading from local?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/truera/trulens/pull/1559/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (sfc-gh-jreini) [inline src/benchmark/trulens/benchmark/benchmark_frameworks/experiments/rag_triad_baselines.ipynb:23]: <pre class="ql-syntax" spellcheck="false">Line #37.                        latencies.append(end_time - start_time)
</pre><p>why do you have to compute latency here? Don't we track this anyways?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/truera/trulens/pull/1559/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

Comment (sfc-gh-jreini) [inline src/benchmark/trulens/benchmark/benchmark_frameworks/experiments/rag_triad_baselines.ipynb:23]: <pre class="ql-syntax" spellcheck="false">Line #46.        write_results(
</pre><p>It'd be better to show the results with trulens-native apis e.g. get_leaderboard rather than writing to csv</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/truera/trulens/pull/1559/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

Comment (sfc-gh-jreini) [inline src/benchmark/trulens/benchmark/benchmark_frameworks/experiments/rag_triad_baselines.ipynb:23]: <pre class="ql-syntax" spellcheck="false">Line #50.        # auc = f_auc(scores)
</pre><p>why removing auc?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/truera/trulens/pull/1559/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (sfc-gh-dhuang) [reply src/benchmark/trulens/benchmark/benchmark_frameworks/experiments/rag_triad_baselines.ipynb:23]: mainly b/c we have F1 score reported already, and AUC is a less meaningful metric for binary classification cases

Comment (sfc-gh-dhuang) [reply src/benchmark/trulens/benchmark/benchmark_frameworks/experiments/rag_triad_baselines.ipynb:23]: this was added before the latency fix came in. keeping it here for reproducibility.

Comment (sfc-gh-dhuang) [reply src/benchmark/trulens/benchmark/benchmark_frameworks/experiments/rag_triad_baselines.ipynb:23]: `ms_marco_hard_neg_balanced` is not available in BEIR. It's a distinct dataset https://huggingface.co/datasets/nixiesearch/ms-marco-hard-negatives

File: src/benchmark/trulens/benchmark/benchmark_frameworks/experiments/competitive_analysis.ipynb [hunk 7]
```diff
@@ -0,0 +1,573 @@
+{
```
Comment (sfc-gh-pdharmana) [inline src/benchmark/trulens/benchmark/benchmark_frameworks/experiments/competitive_analysis.ipynb:1]: <pre class="ql-syntax" spellcheck="false">do you need any sampling ?
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/truera/trulens/pull/1559/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (sfc-gh-pdharmana) [inline src/benchmark/trulens/benchmark/benchmark_frameworks/experiments/competitive_analysis.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #61.                score = 0  # if there is an error, we assume the score is 0
</pre><p>If there is an error, we should skip this row instead of making the score 0? While running these tests, you could hit issues like rate limiting error, and that could impact benchmarks</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/truera/trulens/pull/1559/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (sfc-gh-pdharmana) [inline src/benchmark/trulens/benchmark/benchmark_frameworks/experiments/competitive_analysis.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #68.        write_results(
</pre><p>If we are initializing a trulens app, why write results? Also we are not using tru benchmark experiment? Was that not supposed to help here?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/truera/trulens/pull/1559/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (sfc-gh-pdharmana) [inline src/benchmark/trulens/benchmark/benchmark_frameworks/experiments/competitive_analysis.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #40.    qags_cnn_dm_true_labels = [
</pre><p>These were already computed earlier?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/truera/trulens/pull/1559/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

**Review by sfc-gh-pdharmana — APPROVED** [review]
LGTM with some comments

File: src/benchmark/trulens/benchmark/benchmark_frameworks/experiments/competitive_analysis.ipynb (hunk 7 above)
Comment (sfc-gh-dhuang) [reply src/benchmark/trulens/benchmark/benchmark_frameworks/experiments/competitive_analysis.ipynb:1]: <p><code>qags_xsum_true_labels</code> is derived from <code>qags_xsum</code>,where the sampling already happened     </p>

Comment (sfc-gh-dhuang) [reply src/benchmark/trulens/benchmark/benchmark_frameworks/experiments/competitive_analysis.ipynb:1]: <p>these utils (along with a lot of other stuff) were from the time when I first started running benchmark back in September, and during those periods of time TruBenchmark wrapper was not in a good enough shape to be used. </p><p> </p><p>"not in a good enough shape" here refers to 2 main thing</p><p> </p><ol><li>score generation errors throw off the execution of TruBenchmark app, cuz all scores are generated from one app run </li><li>latency display and casting issue, which was later fixed by Corey's PR <a href="https://github.com/truera/trulens/pull/1522" target="_blank">https://github.com/truera/trulens/pull/1522</a></li></ol>

Comment (sfc-gh-dhuang) [reply src/benchmark/trulens/benchmark/benchmark_frameworks/experiments/competitive_analysis.ipynb:1]: <p>I'll have a separate PR to refactor. This PR is mainly just moving things around  from existing RAG triad benchmark notebook to <code>dataset_preprocessing.py</code></p>
